### ISSUE TITLE
About the source code

### ISSUE THREAD

**Issue by bzhangGo** [zomux/tree2code#1]
Hi @zomux,

We are interested in your paper, and now heading to reproduce the syntactic codes. We'are facing the problem of batching different parsing trees, saving GPU running-time memory, and optimizing the improved semantic hashing module. 
It's so great to know that you are willing to release your source code (as well as running/training script if possible) here. But, is the source code ready for sharing now?

best and thanks,
Biao

**Comment by zomux** [comment]
@bzhangGo Please give me few days to release the code. Sorry for the delay, because I have to write my PhD thesis. I can give you the code before refactoring if you want.

**Comment by bzhangGo** [comment]
thanks. We can wait for the final code. But it would be great if you can share it with us now. Can you send the code to [my email](mailto:b.zhang@ed.ac.uk)?

**Comment by zomux** [comment]
@bzhangGo Hi, all the source code is now available in the repo. Please check it out.

**Comment by bzhangGo** [comment]
thansk.

**Comment by bzhangGo** [comment]
Hi, I have a question about memory usage. 

It seems the current implementation tries to load all the samples into host memory first, and then do iteratively training. Our training corpus is very large, loading all of which perhaps is beyond the capacity of our server.  I notice the option `--opt_limit_datapoints 100000`, which is also suggested in the single gpu training. However, even 800K effective training samples seem too small, compared with the whole training corpus, e.g. 4M in WMT14. 

Any suggestions about how to handle it? thanks.

**Comment by zomux** [comment]
@bzhangGo Can I know how many GPUs do you have in a single machine, and how many sentence pairs in your dataset?

**Comment by bzhangGo** [comment]
I'm using one GPU with around 10M sentence pairs.

**Comment by zomux** [comment]
If you have 8 GPUs, then when you train the model with multi-gpu support, each GPU process only loads 1.2M parse trees into the memory. Then it seems like a viable way. However, I'm still worrying that the CPU memory may not be enough in this case.

And of course, you only have one GPU. Loading the whole parse tree at the beginning will also be very time-consuming.

Let me see whether you can build the tree structures and dump them into many pickles, and load from the pickles in the training time. Actually, the `TreeDataGenerator` class already has an option `cache_path` to allow you load pickled data from cache file.

**Comment by zomux** [comment]
@bzhangGo Let me first check whether you will hit a memory issue. I will get back to you.

**Comment by bzhangGo** [comment]
thanks, @zomux

**Comment by bzhangGo** [comment]
@zomux hi, I met a problem when using one GPU alone.  I'm using the given WMT14 training data and limit the maximum data points to be 100K. Below is my result:
```
[OPTS] Model tag: codebits-8_dtok-wmt14_limit_datapoints-100000_limit_tree_depth-2
Running on 1 GPUs
building trees part 0 ...
[nmtlab] Training TreeAutoEncoder with 74 parameters
[nmtlab] with Adagrad and SimpleScheduler
[nmtlab] Training data has 773 batches
[nmtlab] Running with 1 GPUs (GeForce GTX 1080 Ti)
[valid] loss=6.79 label_accuracy=0.00 * (epoch 1, step 1)
[valid] loss=1.58 label_accuracy=0.52 * (epoch 1, step 194)
[valid] loss=1.55 label_accuracy=0.53 * (epoch 1, step 387)
[valid] loss=1.52 label_accuracy=0.53 * (epoch 1, step 580)
[valid] loss=1.51 label_accuracy=0.53 * (epoch 1, step 773)
[train] loss=1.76 label_accuracy=0.50/s   
[nmtlab] Ending epoch 1, spent 3 minutes  
[valid] loss=1.50 label_accuracy=0.53 * (epoch 2, step 967)
[valid] loss=1.49 label_accuracy=0.54 * (epoch 2, step 1160)
[valid] loss=1.49 label_accuracy=0.54 * (epoch 2, step 1353)
[valid] loss=1.48 label_accuracy=0.53 * (epoch 2, step 1546)
[train] loss=1.52 label_accuracy=0.52/s   
...
[nmtlab] Ending epoch 23, spent 2 minutes  
[valid] loss=1.48 label_accuracy=0.55 (epoch 24, step 17973)
[valid] loss=1.48 label_accuracy=0.55 (epoch 24, step 18166)
[valid] loss=1.47 label_accuracy=0.54 (epoch 24, step 18359)
[valid] loss=1.48 label_accuracy=0.55 (epoch 24, step 18552)
[train] loss=1.23 label_accuracy=0.61h/s  
```
The model fails to converge with one GPU. But, considering the small size of this training setting, I expect I could get much higher accuracy. Can you give me some suggestions when tunning the model with one GPU? thanks.

**Comment by zomux** [comment]
@bzhangGo Hi, `label_accuracy=0.54` is too low.  Please try `--opt_limit_datapoints 800000`,  because the limitation is for each GPU. As you are running on single GPU, you should increase the datapoint limit.

**Comment by zomux** [comment]
@bzhangGo Please post the results after increasing datapoint limit. If not successful, I will try it myself on a single GPU

**Comment by bzhangGo** [comment]
Thanks. Adding more data works,
```
[OPTS] Model tag: codebits-8_dtok-wmt14_limit_datapoints-800000_limit_tree_depth-2
Running on 1 GPUs
building trees part 0 ...
[nmtlab] Training TreeAutoEncoder with 74 parameters
[nmtlab] with Adagrad and SimpleScheduler
[nmtlab] Training data has 6242 batches
[nmtlab] Running with 1 GPUs (GeForce GTX 1080 Ti)
[valid] loss=6.64 label_accuracy=0.00 * (epoch 1, step 1)
[valid] loss=1.06 label_accuracy=0.68 * (epoch 1, step 1561)
[valid] loss=0.73 label_accuracy=0.78 * (epoch 1, step 3121)
[valid] loss=0.64 label_accuracy=0.81 * (epoch 1, step 4681)
[valid] loss=0.59 label_accuracy=0.83 * (epoch 1, step 6241)
[train] loss=0.89 label_accuracy=0.74/s   
[nmtlab] Ending epoch 1, spent 23 minutes  
[valid] loss=0.56 label_accuracy=0.84 * (epoch 2, step 7803)
[valid] loss=0.55 label_accuracy=0.84 * (epoch 2, step 9363)
[valid] loss=0.54 label_accuracy=0.85 * (epoch 2, step 10923)
```

**Comment by Shajiu** [comment]
您好数据集格式大致是个什么情况呢？

**Comment by Shajiu** [comment]
这是最终的源码么？非常感谢

**Comment by zomux** [comment]
@18810979033 Yes, this is the final code.
